### DISCUSSION TITLE
Next.js: Error: ENOENT: no such file or directory

### DISCUSSION THREAD

**Discussion by nbouvrette** [vercel/community#1180] [Help] [answered]
I don't know if this is the right place but before I open a bug I wanted to open a discussion.

I followed the instruction on both:

- https://github.com/vercel/vercel/discussions/5066
- https://vercel.com/docs/concepts/functions/serverless-functions/runtimes#including-additional-files

I used the exact syntax provided on the doc:

```ts
    const file = path.join(process.cwd(), 'emails', 'default.mjml')
    const emailTemplate = readFileSync(file, 'utf8')
```

And I am getting this error on the function logs:

```log
Error: ENOENT: no such file or directory, open '/var/task/emails/default.mjml'
```

I suspect the static analysis is not working because we are using `import { readFileSync } from 'node:fs'` instead of `import { readFileSync } from 'fs'`

Can anyone confirm if this is the case and if I should file a bug?

**Comment by amyegan** [comment]
Thanks for reaching out @nbouvrette. I need to get some more information to help me find what is going wrong.
- Is this a Next.js app? Or are you using a different framework? Please also share the framework version you're using.
- Is this failing on both the server build and local? Or does it work in local dev runs?
- Do you have a public repo or [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) that you can share? That is usually the fastest way for me to see exactly what is going wrong for your app.

**Comment by nbouvrette** [comment]
Hi @amyegan, thanks for answering.

- This is a Next.js app using version `13.0.6`
- It works well locally. The build works on Vercel but the error is triggered when calling the specific API.
- I can try to build a minimal repo but I don't have one available - this might take an hour or two.

As a side note, I also tried the following steps:

- I tried to remove `node:` and it didn't fix the issue
- I tried to use the `vercel.json` configuration with at least 10 different paths to try to fix the same issue that would pop all the time: `Error: The pattern "<15 different paths tried here>" defined in `functions` doesn't match any Serverless Functions.`
- I tried to move my files under `/api` and still didn't fix the issue

I'm basically out of ideas and workaround - it's like if Vercel would ignore `.mjml` files or something?

**Comment by nbouvrette** [comment]
@amyegan ok I was able to reproduce this with a very simple setup - I am suspecting this is a Vercel bug now. Here is the info on the setup:

- Public GitHub repo: https://github.com/nbouvrette/vercel-file-not-found
- Vercel project: https://vercel.com/avansai/vercel-file-not-found 
- Vercel URL: https://vercel-file-not-found-b391gsyq6-avansai.vercel.app/
- If you test it locally:
   - The homepage of the repo returns the content of the mjml file returned by the API
- On Vercel, the API call fails with the same error I had: `Error: ENOENT: no such file or directory, open 'emails/default.mjml'`

Please let me know the next steps as this is blocking our deployments.

Thanks

**Reply by amyegan** [reply]
Thanks for sharing all of this @nbouvrette. I shared it with the Next.js team and asked for their help.

**Reply by nbouvrette** [reply]
Thank you!

**Reply by amyegan** [reply]
They shared this doc with the answer to our problem! The solution is to [use process.cwd()](https://nextjs.org/docs/api-reference/data-fetching/get-static-props#reading-files-use-processcwd) 🙌 

https://nextjs.org/docs/api-reference/data-fetching/get-static-props#reading-files-use-processcwd

**Reply by nbouvrette** [reply]
Hmm I tried that earlier and got the same issue.. let me try it again and circle back.

**Reply by nbouvrette** [reply]
Ok I change the code to replicate the original doc:

```ts
const templateFile = path.join(process.cwd(), "emails", "default.mjml");
const emailTemplate = readFileSync(templateFile, "utf8");
```

It seems like the error is no longer showing up (this should probably be better documented [here](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes#including-additional-files))... I swear I tested this earlier on my other repo, I will try again circle back (probably in 2 hours-ish from now)

**Reply by amyegan** [reply]
This is good feedback, Nicolas. I'm sorry you had to struggle with this. It seems we have some room for improvement in the documentation, and I'm opening a new ticket for that change now.

**Reply by nbouvrette** [reply]
Thanks @amyegan 

But bad news. I wasn't crazy and I did try this before and it didn't work...

I have the following code:

```ts
const emailTemplateFilePath = path.join(process.cwd(), 'emails', 'default.mjml')
const emailTemplate = readFileSync(emailTemplateFilePath, 'utf8')
```

And I am getting the following error still:

```log
2022-12-22T01:09:52.387Z	58ccf936-2e65-4f87-b54d-9c993201fbf4	ERROR	error - unexpected error while open the template file: Error: ENOENT: no such file or directory, open '/var/task/emails/default.mjml'
```

I will try to reproduce it in the example repo, but if I can't it might help to get more details about how those files are "detected" on the Vercel side.

**Reply by nbouvrette** [reply]
Ok, so it looks like the problem is somewhere else.

I copied the [hello.ts](https://github.com/nbouvrette/vercel-file-not-found/blob/main/pages/api/hello.ts) from the example repo, to my private repo. and while it's working on the public repo, it now triggers the error on the private repo.

I suspect that maybe because we are using a Babel plugin, it interferes with the static analysis. I will try to test this theory and circle back

**Reply by nbouvrette** [reply]
@amyegan apparently we had two issues at stake, I would recommend confirming that this can also cause this behavior and if so, adding it to the documentation.

For some reason we had this setting in our Next.js config:

`outputFileTracing: false`

As soon as I removed this, the file was included correctly in the Vercel API. I guess it was a very unlucky set of coincidences on our end but I'm glad I found this even if it took me way longer than I would have hoped.

**Reply by amyegan** [reply]
Oh no! That was a sneaky problem. I'm sorry this took so long to debug, but I'm very happy you found a solution. Thank you for coming back to share the answer. 🙇 

I updated the documentation ticket with this new information.

**Comment by nbouvrette** [comment]
Morning @amyegan,

Unfortunately, I still haven't been able to make it work. I'm using both `process.cwd()` and `outputFileTracing: true` and the `no such file or directory` error remains. The behavior is not consistent with the example repo which seems to work.

At this stage, I tried so many different things that I am running out of ideas. I would love the static analysis to work but until we can find the root cause behind this, maybe I can try to use the [vercel.json](https://vercel.com/docs/concepts/functions/serverless-functions/runtimes#including-additional-files) option in the meantime?

But the problem is when I tried `vercel.json` I keep getting this build error, regardless of the different configs I try:

```
Error: The pattern "api/**/*.ts" defined in `functions` doesn't match any Serverless Functions.
```

I'm using the following config based on the [official doc](https://vercel.com/docs/project-configuration#project-configuration/functions/value-definition):

```json
{
  "functions": {
    "api/**/*.ts": {
      "includeFiles": "emails/**"
    }
  }
}
```

**Reply by amyegan** [reply]
Is your serverless function in a root level `/api` directory, or is it in `/pages/api`? You may need to adjust the `functions` path if you're using Next.js API Routes

**Reply by nbouvrette** [reply]
Yes my bad, I just noticed that and got the build to work. Unfortunately, I am still getting the same (`no such file`) error which makes no sense.

Apparently, this problem is not new - I found this old discussion with similar problems: 

https://github.com/vercel/next.js/discussions/14807

Should we open an official issue on this topic?

Do you have any idea why with 3 different "fixes" (`process.cwd`, `outputFileTracing: false` and `vercel.json`) it's still not working?

Keep in mind all of this works locally too which is even more confusing

**Reply by amyegan** [reply]
Are you able to share a minimal example that more closely matches your real app scenario? The engineers aren't able to reproduce the problem, so it's difficult to find what else might be going wrong

**Reply by nbouvrette** [reply]
I also tried this (still getting the `no such file` error):

```ts
export const config = {
  unstable_includeFiles: ['emails/default.mjml'],
}
```

Also based on this [comment](https://github.com/vercel/next.js/issues/8251#issuecomment-961972409):

> includeFiles in vercel.json isn't used by Next.js.

I haven't seen anywhere this is documented.

Unless I am missing something it looks like including files in Next.js API is not really supported and is being discussed here? https://github.com/vercel/next.js/discussions/32236

**Reply by nbouvrette** [reply]
> Are you able to share a minimal example that more closely matches your real app scenario? The engineers aren't able to reproduce the problem, so it's difficult to find what else might be going wrong

The example repo is as close as I could get to the private repo, and unfortunately, I cannot reproduce it either since I set `outputFileTracing: false` - so it's a big mystery to me. I don't know if your engineers can access my Vercel app that is failing right now?:

https://vercel.com/avansai/avansai-app/Hs3H1buGqTU2Sg4qo4aKqJQjJwNK

I copied the `/api/hello` from the example and it's failing on the private repo.

Otherwise, I would be open to screen share/video call.

**Comment by nbouvrette** [comment, marked as answer]
@amyegan I just found a solution... this seems very hacky - given the [discussion](https://github.com/vercel/next.js/discussions/32236) I shared earlier, it looks like the Next.js team is already aware of this issue but I am not sure they are working on fixing it? Would you know more details?

Here is my solution:

I added a type declaration for my template files in a `d.ts` file in my project:

```ts
declare module '*.mjml' {
  const content: string
  export default content
}
```

Then I added a new Webpack configuration to load these files using the `raw-loader` plugin which just mimics a `readFile` (this is in `next.config.js`):

```js
webpack(config, context) {
  config.module.rules.push(
    {
      test: /\.mjml$/,
      use: 'raw-loader',
    }
  )
  return config
}
```

And now in my API file, I simply import the file so that instead of trying to use the static analysis to guess which files are used, it's just a normal Node dependency:

`import emailTemplate from '../../emails/default.mjml'`

It might be good to have a disclaimer somewhere in the doc that this can be used as a workaround and in some cases, the documented solution just doesn't work... as you can tell from the discussions I'm not the only one with this issue

**Reply by amyegan** [reply]
This sounds like a lot of extra config for a feature that normally just works. 😞 It looks like you're falling into an edge case that the engineers want to solve, but we're having a hard time reproducing the behavior you're seeing. 

If you're able to share a reproduction (or your real site repository if it's public), we would appreciate the opportunity to study it so we can fix the problem. Otherwise, I'm just happy that your site working! Thanks for sharing the solution that worked for you.

**Reply by nbouvrette** [reply]
Unfortunately, I was only able to reproduce it on my private site for now. If the dev team wants, I am happy to jump on an interactive session with them to help them nail this issue once and for all.

Otherwise, there might be other good examples available on the [discussion](https://github.com/vercel/next.js/discussions/32236) I shared. I didn't have time to check them since I was more worried about finding a fix!

Thanks for all your help
